com.ca.lsp.cobol/lsp-core-cobol-parser/src/main/java/com/ca/lsp/core/cobol/visitor/CobolVisitor.java

```diff
@@ -38,212 +48,293 @@
 
-  private void getSemanticError(
-      String wrongToken, int startLine, int charPositionInLine, String correctWord) {
-    errors.add(
-        SyntaxError.syntaxError()
-            .position(
-                new Position(
-                    documentUri,
-                    charPositionInLine,
-                    getWrongTokenStopPosition(wrongToken, charPositionInLine),
-                    startLine,
-                    charPositionInLine))
-            .suggestion("A misspelled word, maybe you want to put " + correctWord)
-            .severity(WARNING_LEVEL)
-            .build());
+  @Override
+  public Class visitErrorNode(ErrorNode node) {
+    Token token = node.getSymbol();
+    Position position = calculatePosition(token);
+    mapping.put(token, position);
+
+    return super.visitTerminal(node);
   }
 
-  private boolean checkForVariable(
-      String variable, int startLine, int charPositionInLine, ParserRuleContext ctx) {
-    if (!semanticContext.getVariables().contains(variable)) {
-      throwSuggestion(variable, startLine, charPositionInLine);
-      return false;
-    } else if (ctx instanceof QualifiedDataNameFormat1Context
-        && ((QualifiedDataNameFormat1Context) ctx).qualifiedInData() != null) {
-      iterateOverQualifiedDataNames(
-          (QualifiedDataNameFormat1Context) ctx, variable, startLine, charPositionInLine);
-      addUsage(semanticContext.getVariables(), variable, ctx);
+  private void checkForVariable(String variable, QualifiedDataNameFormat1Context ctx) {
+    checkVariableDefinition(variable, ctx.getStart());
+    addUsage(variables, variable, ctx.getStart());
+
+    if (ctx.qualifiedInData() != null) {
+      iterateOverQualifiedDataNames(ctx, variable);
     }
-    return true;
   }
 
-  private void iterateOverQualifiedDataNames(
-      QualifiedDataNameFormat1Context ctx, String variable, int startLine, int charPositionInLine) {
+  private void iterateOverQualifiedDataNames(QualifiedDataNameFormat1Context ctx, String variable) {
+    String child = variable;
+    Token childToken = ctx.getStart();
     for (QualifiedInDataContext node : ctx.qualifiedInData()) {
-      if (node.inData() != null) {
-        visitInData(variable, startLine, charPositionInLine, node.inData());
-        DataName2Context context = node.inData().dataName2();
-        variable = context.getText();
-        addUsage(semanticContext.getVariables(), context);
-      } else {
-        visitInTable(variable, startLine, charPositionInLine, node.inTable());
-        DataName2Context context = node.inTable().tableCall().dataName2();
-        variable = context.getText();
-        addUsage(semanticContext.getVariables(), context);
-      }
+
+      DataName2Context context = getDataName2Context(node);
+      String parent = context.getText().toUpperCase();
+      Token parentToken = context.getStart();
+
+      checkVariableDefinition(parent, parentToken);
+      checkVariableStructure(parent, child, childToken);
```

Comment:
This may throw similar semantic errors for different reasons. To be fixed in #321